code/datums/components/puzzgrid.dm

```diff
@@ -0,0 +1,272 @@
+	RegisterSignal(parent, COMSIG_ATOM_ATTACK_HAND, .proc/on_attack_hand)
+
+/datum/component/puzzgrid/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_ATOM_ATTACK_HAND)
+
+/datum/component/puzzgrid/proc/on_attack_hand(atom/source, mob/user)
+	SIGNAL_HANDLER
+
+	INVOKE_ASYNC(src, .proc/ui_interact, user)
+
+/datum/component/puzzgrid/ui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Puzzgrid")
+		ui.open()
+
+/datum/component/puzzgrid/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if (.)
+		return .
+
+	switch (action)
+		if ("select")
+			return try_select(params["answer"])
+		if ("unselect")
+			return try_unselect(params["answer"])
+
+	return TRUE
+
+/datum/component/puzzgrid/proc/try_select(answer)
+	if (!(answer in all_answers))
+		return FALSE
+
+	if (!COOLDOWN_FINISHED(src, wrong_group_select_cooldown))
+		return TRUE
+
+	selected_answers |= answer
+
+	if (selected_answers.len < PUZZGRID_CELL_COUNT)
+		return TRUE
+
+	var/list/current_selected_answers = selected_answers
+	selected_answers = list()
+
+	search_group:
+		for (var/datum/puzzgrid_group/puzzgrid_group in (puzzgrid.groups - solved_groups))
+			for (var/selected_answer in current_selected_answers)
+				if (!(selected_answer in puzzgrid_group.answers))
+					continue search_group
+
+			// This group has the right answers
+			solved_groups += puzzgrid_group
+
+			if (solved_groups.len == puzzgrid.groups.len - 1)
+				on_victory()
+			else
+				update_static_data_for_all_viewers()
+
+			return TRUE
+
+	COOLDOWN_START(src, wrong_group_select_cooldown, 0.2 SECONDS)
+
+	if (solved_groups.len == puzzgrid.groups.len - 2)
+		lives -= 1
+
+		if (lives == 0)
+			out_of_lives()
+
+	return TRUE
+
+/datum/component/puzzgrid/proc/try_unselect(answer)
+	selected_answers -= answer
+	return TRUE
+
+/datum/component/puzzgrid/proc/on_victory()
+	report_answers()
+	on_victory_callback?.InvokeAsync()
+	qdel(src)
+
+/datum/component/puzzgrid/proc/out_of_lives()
+	var/atom/movable/movable_parent = parent
+	if (istype(movable_parent))
+		movable_parent.say("Ran out of lives!")
+
+	fail()
+
+/datum/component/puzzgrid/proc/out_of_time()
+	var/atom/movable/movable_parent = parent
+	if (istype(movable_parent))
+		movable_parent.say("Ran out of time!")
+
+	fail()
+
+/datum/component/puzzgrid/proc/fail()
+	report_answers()
+	on_fail_callback?.InvokeAsync()
+	qdel(src)
+
+/datum/component/puzzgrid/proc/report_answers()
+	var/atom/atom_parent = parent
+
+	var/list/answers = list()
+	for (var/datum/puzzgrid_group/puzzgrid_group as anything in puzzgrid.groups)
+		var/list/answers_encoded = list()
+		for (var/answer in puzzgrid_group.answers)
+			answers_encoded += html_encode(answer)
+
+		answers += span_boldnotice("<p>[answers_encoded.Join(", ")]</p>") + span_notice("<p>[html_encode(puzzgrid_group.description)]</p>")
+
+	atom_parent.audible_message(answers.Join("<p>-----</p>"))
```

Comment:
Audible seems like weird choice here

Reply:
Do we have a good way to just show a message to people around you regardless of being blind/deaf, show_message didn't look like it

Reply:
not really just direct to_chat

Reply:
Shame, oh well I'll just loop viewers or something